Use more smart pointers in SubresourceLoader

fujii opened the pull request:
#### 85c7594d6be4e0c79e0981f2f6638baddd6dac5e
<pre>
Use more smart pointers in SubresourceLoader
<a href="https://bugs.webkit.org/show_bug.cgi?id=263999">https://bugs.webkit.org/show_bug.cgi?id=263999</a>

Reviewed by Chris Dumez.

* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::protectedDocumentLoader const):
* Source/WebCore/loader/ResourceLoader.h:
(WebCore::ResourceLoader::protectedCachedResource const):
* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::SubresourceLoader::init):
(WebCore::SubresourceLoader::willSendRequestInternal):
(WebCore::SubresourceLoader::didSendData):
(WebCore::SubresourceLoader::didReceivePreviewResponse):
(WebCore::SubresourceLoader::didReceiveResponse):
(WebCore::SubresourceLoader::didReceiveBuffer):
(WebCore::SubresourceLoader::responseHasHTTPStatusCodeError const):
(WebCore::SubresourceLoader::checkRedirectionCrossOriginAccessControl):
(WebCore::SubresourceLoader::didFinishLoading):
(WebCore::SubresourceLoader::didFail):
(WebCore::SubresourceLoader::willCancel):
(WebCore::SubresourceLoader::didCancel):
(WebCore::SubresourceLoader::notifyDone):
(WebCore::SubresourceLoader::releaseResources):
(WebCore::SubresourceLoader::reportResourceTiming):

Canonical link: <a href="https://commits.webkit.org/270049@main">https://commits.webkit.org/270049@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/afc8527affea36736ec18ac8687dc48fe0d25a57

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24390 "check-webkit-style (failure)") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2586 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25503 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26521 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22438 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24661 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4220 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/57 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22863 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24636 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2010 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21069 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27107 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1748 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21995 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22234 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22310 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1685 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/32 "Found 2 new test failures: fast/replaced/pdf-as-embed-with-no-mime-type-is-not-blank.html, fast/scrolling/scroll-to-focused-element-asynchronously.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/2971 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2099 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3114 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2054 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -181,14 +181,15 @@ void SubresourceLoader::init(ResourceRequest&& request, CompletionHandler<void(b
         if (!initialized)
             return completionHandler(false);
-        if (!m_documentLoader) {
+        auto documentLoader = protectedDocumentLoader();
```
The preferred pattern is:
```
RefPtr documentLoader = m_documentLoader;
```
or
```
RefPtr documentLoader = this->documentLoader();
```

We prefer not to use `auto` to hide smart pointers anymore as we want to make it clear we're using smart pointers and the code is safe.

fujii replied:
Do you think we need `protected` prefix here? https://webkit.org/code-style-guidelines/#names-protectors

cdumez replied:
No, protectors are no longer a thing with our new smart pointer idioms. This is just a plain local variable. And protectedThis protectors are no longer useful since we now expect the call sites to protect the object they're calling a member function on.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -294,9 +295,9 @@ void SubresourceLoader::willSendRequestInternal(ResourceRequest&& newRequest, co
 
         auto originalOrigin = SecurityOrigin::create(redirectResponse.url());
-        m_site = m_documentLoader->cachedResourceLoader().computeFetchMetadataSite(newRequest, m_resource->type(), options().mode, originalOrigin, m_site);
+        m_site = protectedDocumentLoader()->cachedResourceLoader().computeFetchMetadataSite(newRequest, m_resource->type(), options().mode, originalOrigin, m_site);
```
Seems we should cache the DocumentLoader in a RefPtr local variable since it is used at least twice. Avoids refunding churn.

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -394,7 +395,7 @@ void SubresourceLoader::didReceiveResponse(const ResourceResponse& response, Com
     // Implementing step 10 of https://fetch.spec.whatwg.org/#main-fetch for service worker responses.
     if (response.source() == ResourceResponse::Source::ServiceWorker && response.url() != request().url()) {
-        auto& loader = m_documentLoader->cachedResourceLoader();
+        auto& loader = protectedDocumentLoader()->cachedResourceLoader();
```
`Ref loader = `

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -423,7 +424,8 @@ void SubresourceLoader::didReceiveResponse(const ResourceResponse& response, Com
         response.includeCertificateInfo();
 
-    if (!m_resource) {
+    auto resource = protectedCachedResource();
```
`CachedResourceHandler resource = m_resource.get();`

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -499,19 +501,20 @@ void SubresourceLoader::didReceiveResponse(const ResourceResponse& response, Com
         if (reachedTerminalState())
             return;
 
+        auto protectedResource = protectedCachedResource();
```
ditto.

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -499,19 +501,20 @@ void SubresourceLoader::didReceiveResponse(const ResourceResponse& response, Com
     bool isResponseMultipart = response.isMultipart();
     if (options().mode != FetchOptions::Mode::Navigate)
-        LinkLoader::loadLinksFromHeader(response.httpHeaderField(HTTPHeaderName::Link), m_documentLoader->url(), *m_frame->document(), LinkLoader::MediaAttributeCheck::SkipMediaAttributeCheck);
+        LinkLoader::loadLinksFromHeader(response.httpHeaderField(HTTPHeaderName::Link), protectedDocumentLoader()->url(), *m_frame->document(), LinkLoader::MediaAttributeCheck::SkipMediaAttributeCheck);
```
`*m_frame->protectedDocument()`

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -550,12 +553,13 @@ void SubresourceLoader::didReceiveBuffer(const FragmentedSharedBuffer& buffer, l
 #endif
 
-    ASSERT(m_resource);
+    auto resource = protectedCachedResource();
```
ditto.

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -565,15 +569,16 @@ void SubresourceLoader::didReceiveBuffer(const FragmentedSharedBuffer& buffer, l
 bool SubresourceLoader::responseHasHTTPStatusCodeError() const
 {
-    if (m_resource->response().httpStatusCode() < 400 || m_resource->shouldIgnoreHTTPStatusCodeErrors())
+    auto resource = protectedCachedResource();
```
ditto

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -633,7 +638,7 @@ static void logResourceLoaded(LocalFrame* frame, CachedResource::Type type)
 Expected<void, String> SubresourceLoader::checkResponseCrossOriginAccessControl(const ResourceResponse& response)
 {
-    if (!m_resource->isCrossOrigin() || options().mode != FetchOptions::Mode::Cors)
+    if (!protectedCachedResource()->isCrossOrigin() || options().mode != FetchOptions::Mode::Cors)
```
I think it is probably OK because it is not hot code but you don't really need to ref to call trivial getters.

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -789,25 +795,25 @@ void SubresourceLoader::didFail(const ResourceError& error)
     ASSERT(!reachedTerminalState());
-    CachedResourceHandle protectedResource { m_resource.get() };
-    LOG(ResourceLoading, "Failed to load '%s'.\n", m_resource->url().string().latin1().data());
+    auto protectedResource = protectedCachedResource();
```
ditto

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -789,25 +795,25 @@ void SubresourceLoader::didFail(const ResourceError& error)
 
-    if (m_frame->document() && error.isAccessControl() && error.domain() != InspectorNetworkAgent::errorDomain() && m_resource->type() != CachedResource::Type::Ping)
+    if (m_frame->document() && error.isAccessControl() && error.domain() != InspectorNetworkAgent::errorDomain() && protectedResource->type() != CachedResource::Type::Ping)
         m_frame->document()->addConsoleMessage(MessageSource::Security, MessageLevel::Error, error.localizedDescription());
```
`m_frame->protectedDocument()`

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -789,25 +795,25 @@ void SubresourceLoader::didFail(const ResourceError& error)
 
     Ref<SubresourceLoader> protectedThis(*this);
-    CachedResourceHandle<CachedResource> protectResource(m_resource.get());
+    CachedResourceHandle<CachedResource> protectResource(protectedResource.get());
```
Seems we should drop this variable?

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -789,25 +795,25 @@ void SubresourceLoader::didFail(const ResourceError& error)
 
-    if (m_resource->resourceToRevalidate())
+    if (protectedResource->resourceToRevalidate())
         MemoryCache::singleton().revalidationFailed(*m_resource);
```
Should use the protectedResource

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -841,7 +847,7 @@ void SubresourceLoader::willCancel(const ResourceError& error)
         memoryCache.revalidationFailed(*m_resource);
-    m_resource->setResourceError(error);
+    protectedCachedResource()->setResourceError(error);
     memoryCache.remove(*m_resource);
```
Should use the protectedResource.

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -850,12 +856,13 @@ void SubresourceLoader::didCancel(const ResourceError&)
         return;
 
-    ASSERT(m_resource);
+    auto protectedResource = protectedCachedResource();
```
ditto.

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -870,15 +877,15 @@ void SubresourceLoader::notifyDone(LoadCompletionType type)
 #endif
-    if (m_documentLoader)
-        m_documentLoader->cachedResourceLoader().loadDone(type, shouldPerformPostLoadActions);
+    if (auto documentLoader = protectedDocumentLoader())
```
ditto.

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -870,15 +877,15 @@ void SubresourceLoader::notifyDone(LoadCompletionType type)
         return;
-    if (m_documentLoader)
-        m_documentLoader->removeSubresourceLoader(type, this);
+    if (auto documentLoader = protectedDocumentLoader())
```
ditto.

fujii replied:
Fixed.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -903,7 +910,7 @@ void SubresourceLoader::reportResourceTiming(const NetworkLoadMetrics& networkLo
         return;
 
-    Document* document = m_documentLoader->cachedResourceLoader().document();
+    Document* document = protectedDocumentLoader()->cachedResourceLoader().document();
```
`RefPtr document`

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -921,7 +928,7 @@ void SubresourceLoader::reportResourceTiming(const NetworkLoadMetrics& networkLo
 
     ASSERT(options().initiatorContext == InitiatorContext::Document);
-    m_documentLoader->cachedResourceLoader().resourceTimingInformation().addResourceTiming(*m_resource, *document, WTFMove(resourceTiming));
+    protectedDocumentLoader()->cachedResourceLoader().resourceTimingInformation().addResourceTiming(*m_resource, *document, WTFMove(resourceTiming));
```
Should use `*protectedResource()`

fujii replied:
Fixed.

cdumez reviewed (changes requested):
Only partial adoption and doesn't quite follow our usual patterns.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -565,15 +570,16 @@ void SubresourceLoader::didReceiveBuffer(const FragmentedSharedBuffer& buffer, l
 bool SubresourceLoader::responseHasHTTPStatusCodeError() const
 {
-    if (m_resource->response().httpStatusCode() < 400 || m_resource->shouldIgnoreHTTPStatusCodeErrors())
+    CachedResourceHandle protectedResource = m_resource.get();
```
I think this should just be called `resource`.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -729,14 +736,14 @@ void SubresourceLoader::didFinishLoading(const NetworkLoadMetrics& networkLoadMe
         return;
     ASSERT(!reachedTerminalState());
-    ASSERT(!m_resource->resourceToRevalidate());
+    CachedResourceHandle protectedResource = m_resource.get();
```
just `resource`

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -789,25 +796,24 @@ void SubresourceLoader::didFail(const ResourceError& error)
     ASSERT(!reachedTerminalState());
-    CachedResourceHandle protectedResource { m_resource.get() };
-    LOG(ResourceLoading, "Failed to load '%s'.\n", m_resource->url().string().latin1().data());
+    CachedResourceHandle protectedResource = m_resource.get();
```
just `resource`

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -833,29 +839,31 @@ void SubresourceLoader::willCancel(const ResourceError& error)
     LOG(ResourceLoading, "Cancelled load of '%s'.\n", m_resource->url().string().latin1().data());
 
     Ref<SubresourceLoader> protectedThis(*this);
+    CachedResourceHandle protectedResource = m_resource.get();
```
just `resource`

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -833,29 +839,31 @@ void SubresourceLoader::willCancel(const ResourceError& error)
         return;
 
-    ASSERT(m_resource);
+    CachedResourceHandle protectedResource = m_resource.get();
```
just `resource`

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -892,36 +900,37 @@ void SubresourceLoader::releaseResources()
 {
-    ASSERT(m_resource);
-    if (!m_resource || !ResourceTimingInformation::shouldAddResourceTiming(*m_resource))
+    CachedResourceHandle protectedResource = m_resource.get();
```
just `resource`

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -423,7 +425,8 @@ void SubresourceLoader::didReceiveResponse(const ResourceResponse& response, Com
         response.includeCertificateInfo();
 
-    if (!m_resource) {
+    CachedResourceHandle protectedResource = m_resource.get();
```
just `resource`, not `protectedResource`. That's the old style.

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -491,27 +494,28 @@ void SubresourceLoader::didReceiveResponse(const ResourceResponse& response, Com
         if (reachedTerminalState())
             return;
 
+        CachedResourceHandle protectedResource = m_resource.get();
```
Just `resource`

cdumez commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -550,12 +554,13 @@ void SubresourceLoader::didReceiveBuffer(const FragmentedSharedBuffer& buffer, l
 #endif
 
-    ASSERT(m_resource);
+    CachedResourceHandle protectedResource = m_resource.get();
```
just `resource`

cdumez reviewed (approved):
LGTM but please address all comments before landing.

fujii committed:
Use more smart pointers in SubresourceLoader
https://bugs.webkit.org/show_bug.cgi?id=263999

Reviewed by Chris Dumez.

* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::protectedDocumentLoader const):
* Source/WebCore/loader/ResourceLoader.h:
(WebCore::ResourceLoader::protectedCachedResource const):
* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::SubresourceLoader::init):
(WebCore::SubresourceLoader::willSendRequestInternal):
(WebCore::SubresourceLoader::didSendData):
(WebCore::SubresourceLoader::didReceivePreviewResponse):
(WebCore::SubresourceLoader::didReceiveResponse):
(WebCore::SubresourceLoader::didReceiveBuffer):
(WebCore::SubresourceLoader::responseHasHTTPStatusCodeError const):
(WebCore::SubresourceLoader::checkRedirectionCrossOriginAccessControl):
(WebCore::SubresourceLoader::didFinishLoading):
(WebCore::SubresourceLoader::didFail):
(WebCore::SubresourceLoader::willCancel):
(WebCore::SubresourceLoader::didCancel):
(WebCore::SubresourceLoader::notifyDone):
(WebCore::SubresourceLoader::releaseResources):
(WebCore::SubresourceLoader::reportResourceTiming):

Canonical link: https://commits.webkit.org/270049@main
